addon/components/paper-input.js

```diff
@@ -3,61 +3,74 @@ import BaseFocusable from './base-focusable';
+  classNameBindings: [
+    'hasValue:md-input-has-value',
+    'focus:md-input-focused',
+    'isInvalid:md-input-invalid',
+    'iconFloat:md-icon-float',
+    'icon:md-has-icon'
+  ],
   type: 'text',
   autofocus: false,
   tabindex: -1,
   hideAllMessages: false,
-  hasValue: Ember.computed.notEmpty('value'),
+  isTouched: false,
+
+  hasValue: computed.notEmpty('value'),
 
-  inputElementId: Ember.computed('elementId', function() {
+  inputElementId: computed('elementId', function() {
     return `input-${this.get('elementId')}`;
   }),
 
-  isInvalid: Ember.computed('isTouched', 'value', function() {
-    return this.validate();
+  isInvalid: computed('isTouched', 'value', function() {
+    return this.get('isTouched') && isPresent(this.get('errorMessages'));
   }),
 
-  renderCharCount: Ember.computed('value', function() {
+  renderCharCount: computed('value', function() {
     let currentLength = this.get('value') ? this.get('value').length : 0;
     return `${currentLength}/${this.get('maxlength')}`;
   }),
 
-  iconFloat: Ember.computed.and('icon', 'label'),
+  iconFloat: computed.and('icon', 'label'),
 
   didInsertElement() {
     if (this.get('textarea')) {
-      this.setupTextarea();
-    }
-  },
-
-  setupTextarea() {
-    let textarea = this.$().children('textarea').first();
-    let textareaNode = textarea.get(0);
-    let container = this.get('element');
-    let minRows = NaN;
-    let lineHeight = null;
+      let textarea = this.$().children('textarea').first();
+      let textareaNode = textarea.get(0);
+      let container = this.get('element');
+      let minRows = NaN;
+      let lineHeight = null;
+
+      if (textareaNode.hasAttribute('rows')) {
+        minRows = parseInt(textareaNode.getAttribute('rows'));
+      }
 
-    if (textareaNode.hasAttribute('rows')) {
-      minRows = parseInt(textareaNode.getAttribute('rows'));
-    }
+      textarea.on('keydown input', () => {
+        this.growTextarea(textarea, textareaNode, container, minRows, lineHeight);
+      });
 
-    textarea.on('keydown input', () => {
-      this.growTextarea(textarea, textareaNode, container, minRows, lineHeight);
-    });
+      if (isNaN(minRows)) {
+        textarea.attr('rows', '1');
 
-    if (isNaN(minRows)) {
-      textarea.attr('rows', '1');
+        textarea.on('scroll', () => {
+          this.onScroll(textareaNode);
+        });
+      }
 
-      textarea.on('scroll', () => {
-        this.onScroll(textareaNode);
-      });
+      $(window).on('resize', this.growTextarea(textarea, textareaNode, container, minRows, lineHeight));
     }
+  },
 
-    Ember.$(window).on('resize', this.growTextarea(textarea, textareaNode, container, minRows, lineHeight));
+  willDestroyElement() {
+    if (this.get('textarea')) {
+      $(window).off('resize', this.growTextarea);
+      this.$().children('textarea').first().off('keydown input scroll');
```

Comment:
And when removing event handlers as well.
The reason for this is to make sure we don't remove unwanted event handlers. e.g another component may attach a handler to `window`, the user may be using something that attaches to `textarea`.
By using off with `event.someString` we make sure we target that specific handler.

More info: http://api.jquery.com/event.namespace/